### PR TITLE
Pages Plugins

### PR DESCRIPTION
**Pull request by GregBrimble**
This allows people to compile Plugins for Pages Functions. A compiled plugin is a factory for something incredibly similar to a compiled Pages Functions worker, but with just a couple of [minor tweaks](https://github.com/gregbrimble/pages-plugins#changes).

### CODE REVIEW

**Comment by threepointone** [comment]
Is this urgent? Could we please table this till I'm back? Thank you!

cc @geelen

File: .changeset/spicy-knives-rush.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+"wrangler": patch
+---
+
+Adds a `--plugin` option to `wrangler pages functions build` which compiles a Pages Plugin. More information about these WIP plugins can be found here: https://github.com/gregbrimble/pages-plugins. This wrangler build is required for both the development of, and inclusion of, plugins.
```
Comment (jbergstroem) [inline .changeset/spicy-knives-rush.md:5]:
```suggestion
Adds a `--plugin` option to `wrangler pages functions build` which compiles a Pages Plugin. More information about these WIP plugins can be found here: https://github.com/GregBrimble/pages-plugins. This wrangler build is required for both the development of, and inclusion of, plugins.
```

(the current url 404ed for me)

File: packages/wrangler/pages/functions/template-plugin.ts [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+  const onRequest: PagesPluginFunction = async (workerContext) => {
+    let { request } = workerContext;
+    const { env, next, data } = workerContext;
+
+    const url = new URL(request.url);
+    const relativePathname =
+      url.pathname.split(workerContext.functionPath)[1] || "/";
+
+    const handlerIterator = executeRequest(request, relativePathname);
+    const pluginNext = async (input?: RequestInfo, init?: RequestInit) => {
```
Comment (WalshyDev) [inline packages/wrangler/pages/functions/template-plugin.ts:110]: How come you're making a function for this? You call it immediately anyway

**Review by WalshyDev — COMMENTED** [review]
Apologies for the delay, looks good, few small comments.

Also wonder if we should just have a `plugins` subcommand instead. Then we can have `wrangler pages plugins build` and even things like `wrangler pages plugins install`

File: packages/wrangler/pages/functions/template-plugin.ts [hunk 3]
```diff
@@ -0,0 +1,146 @@
+import { match } from "path-to-regexp";
+import type { HTTPMethod } from "./routes";
+
+/* TODO: Grab these from @cloudflare/workers-types instead */
```
[the remark this answers is not in the document]
Comment (GregBrimble) [reply packages/wrangler/pages/functions/template-plugin.ts:4]: Before this goes out, for sure. People will want the typings for `pluginArgs` etc.

File: packages/wrangler/pages/functions/template-plugin.ts (hunk 2 above)
Comment (GregBrimble) [reply packages/wrangler/pages/functions/template-plugin.ts:110]: It's called recursively: https://github.com/cloudflare/wrangler2/pull/349/files/ebc66999270ab87a65c460b55c27919e5337d8fb#diff-981b507bdc26d5a12df998ad9e1d49b601c6cc4cd542d253fa85322b6b430c5dR122

**Pull request merged**